Update readme

RLin8103 opened the pull request:
Pull Request: Update README

Overview
This pull request updates the README document to provide a more comprehensive and detailed guide on the RetNet vs Transformers project. The revisions include an expanded introduction, detailed installation and setup instructions, and a thorough explanation of the project's features and functionalities. The aim is to make the project more accessible to new contributors and users, ensuring that they have all the information they need to get started.

Key Updates
Project Introduction: Clarified the comparison and analysis focus between RetNet and Transformers, utilizing Microsoft's TorchScale.
Reference to Original Paper: Added a reference to the foundational paper that inspires this project, enhancing academic credibility.
Installation and Setup: Detailed steps for cloning the repository, setting up the Python environment with Mamba, and preparing the environment for running the project's scripts.
Features Explanation: Elaborated on the Grid Search feature, Hugging Face and PyTorch Lightning integration, and TorchScale's role in the project. This section now provides a clearer view of the project's capabilities and its technical advancements.
Execution Workflow: Outlined the steps to prepare and execute the project, including data download, tokenizer training, data tokenization, and model running instructions.
Benchmarking and Results: Mentioned the use of EleutherAI's language model evaluation harness for empirical evaluation, linking to the project's research findings.
Acknowledgments and Citations: Updated to recognize contributions from individuals and organizations, and provided citations for referenced works.

Reason for Changes
The README update aims to streamline the project onboarding process, provide clear guidance on how to utilize the project's tools, and offer insights into the project's research and development focus. By enhancing the documentation, we hope to facilitate a better understanding and engagement with the project for developers, researchers, and enthusiasts alike.

Request for Review
Feedbacks are welcomed on the updated README for any further clarifications or additions that might be necessary.

RLin8103 committed:
Add Skeleton Structure to README.md

RLin8103 committed:
Fix Format Issues

RLin8103 committed:
Fix Code Format Issue in Readme

RLin8103 committed:
Add More Titles to the Readme Skeleton

RLin8103 committed:
Add Python Version Requirements to Installation and Setup

RLin8103 committed:
Add Mamba Instruction to Setup

RLin8103 committed:
Fix Citation Format

RLin8103 committed:
Fix Retnet Paper Link and Citation Version

RLin8103 committed:
Update Retnet Paper Link

RLin8103 committed:
Add Preparing the Environment and Execution Workflow Steps

RLin8103 committed:
Add Link to Paper and Clean Up Some Extra Information

RLin8103 committed:
Add Code Instruction for Execution Workflow Section

RLin8103 committed:
Add Paper Link to Results Section and Add Information to Acknowledgements Section

RLin8103 committed:
Fix Formatting

RLin8103 committed:
Add Grid Search Feature Description

RLin8103 committed:
Add Hugging Face Feature Description and Fix Formatting Issues

RLin8103 committed:
Add Lightning Feature Description

RLin8103 committed:
Add Torchscale Feature Description

RLin8103 committed:
Fix Some Wording

RLin8103 committed:
Fix Some Wording of TorchScale Feature Description

DrewGalbraith committed:
Update README.md

Break the link to overleaf all the way

DrewGalbraith committed:
Update README.md

Correct title of paper cited

DrewGalbraith committed:
Update README.md

Punctuation fixes and command unification

DrewGalbraith committed:
Update README.md with nvidia command and correct grid_search parameters

DrewGalbraith committed:
Make README.md a little prettier

DrewGalbraith committed:
Update README.md with correct grid_search parameters

DrewGalbraith committed:
Update README.md

I added the number of unique models generated by our grid search data_run.

DrewGalbraith committed:
Delete unecessary line from nvidia section.

JayOrten committed:
add section on yaml

JayOrten committed:
add setup to execution

DrewGalbraith committed:
Upload a PDF version of Retnet paper draft 

Upload a PDF version of Retnet paper draft to link into read me while we figure out Overleaf non-editable link sharing.

DrewGalbraith committed:
Update README.md with pdf link to paper draft

DrewGalbraith committed:
Update README.md with lite benchmark info

DrewGalbraith committed:
Alphabetize citations in README.md

DrewGalbraith committed:
Restructure acknowledgements in README.md

DrewGalbraith committed:
Clarify benchmarks section of README.md

RLin8103 commented:
@nprisbrey, @JayOrten, @JacksonSearle, @KimballJardine, and @calvin-seamons, you can review this pull request.

DrewGalbraith reviewed (commented):
I think these updates are clean and very clear. There will certainly be some work be done later on this, especially regarding benchmarks, but it looks good enough to me right now. It's certainly more informative than what we have now.

DrewGalbraith commented on README.md:
```diff
@@ -1,116 +1,268 @@
+
+We use EleutherAI's open-source language model evaluation harness to empirically evaluate our models across a suite of different standardized tasks. These are largely 2-5 option multiple choice tasks, thus a randomly initialized language model will score 50-20% randomly guessing.
+
+You can find more information in the Results section of our paper [Are You Using Retentive Networks?](https://www.overleaf.com/read/fgqvqnmzfncf#a32eb5).
```
This link needs to be removed. Is there an easy way to eliminate it from our git history entirely?

RLin8103 replied:
Oops, that is my bad.

DrewGalbraith commented on README.md:
```diff
@@ -1,116 +1,268 @@
+
+## Citations
+
+This can also be found on the citation section or our paper [Are You Using Retentive Networks?](https://www.overleaf.com/read/fgqvqnmzfncf#a32eb5).
```
This link also needs to be removed, ideally from our git history entirely.

DrewGalbraith reviewed (changes requested):
We have two instances of the Overleaf link in this file. Because this file grants read/write permission to anyone with the link, I think it's a liability to be sharing it in an open repo. We should remove it from the branch before merging and ideally from the git history entirely, if it isn't too difficult.

DrewGalbraith committed:
Update README.md with agnostic mamba environement name

DrewGalbraith committed:
Update README.md titlt to refelct project title

Most popular projects will use the project/paper title as their README.md title. (See BigBench, BBH, ElutherAI LM_eval, etc.)

DrewGalbraith committed:
capitalize mamba placeholder to <YOUR_ENV_NAME>

nprisbrey commented on README.md:
```diff
@@ -1,116 +1,268 @@
+
+**Code Overview:**
+
+- **`RetNetModelHF`**: Implements the RetNet architecture as a subclass of `PreTrainedModel`, utilizing Hugging Face's utilities and standards for model configuration, serialization, and compatibility with the Transformers library.
+- **`TransformerModelHF`**: Adapts the traditional Transformer architecture within the Hugging Face framework, following similar principles as the RetNet model for easy integration and use.
```
Why is the wording for these two classes different? Don't they implement the TorchScale code in the same way? @DrewGalbraith

tristancorncroissant replied:
I'm not sure and yes, they do.

RLin8103 replied:
What is your suggestion on fixing this @nprisbrey?

nprisbrey replied:
I would think that since this code is almost exactly the same, a different description is misleading because it implies that the code is different. I would think something like the following would be better, though once the descriptions are the same, it aids us in arriving to the questions once again of "Is this even needed in the README? Why isn't this just a doc string instead?"

**Code Overview:**

- **`RetNetModelHF`**: Implements the RetNet architecture as a subclass of `PreTrainedModel`, using Hugging Face's utilities and standards for model configuration, serialization, and compatibility with the Transformers library.
- **`TransformerModelHF`**: Implements the Transformer architecture as a subclass of `PreTrainedModel`, using Hugging Face's utilities and standards for model configuration, serialization, and compatibility with the Transformers library.

RLin8103 replied:
I changed it to your description @nprisbrey.

nprisbrey committed:
Change Wording Of Some Sections

nprisbrey committed:
Fix Relative Links

nprisbrey commented on README.md:
````diff
@@ -1,116 +1,267 @@
+If you don't have Mamba installed, you can follow the [Mamba documentation](https://mamba.readthedocs.io/en/latest/installation/mamba-installation.html) to get set up. Once Mamba is installed, create a new environment with Python 3.11:
 
-# Training loop
-for epoch in range(num_epochs):
-    # Training process
-    ...
-    scheduler.step()  # Update learning rate
+```bash
+# Create a new Mamba environment with Python 3.11 and specify a name
+mamba create -n <YOUR_ENV_NAME> python=3.11
+
+# Activate the Mamba environment
+mamba activate <YOUR_ENV_NAME>
````
Is Mamba even required? I use the repository locally without Mamba. I think that it would be useful to mention that Mamba is recommended for HPC environments and that it is **optional**, not required.

RLin8103 replied:
Should we delete this part or do we should just say that it is not required @nprisbrey?

tristancorncroissant replied:
Marked this as optional in this [commit](https://github.com/DRAGNLabs/301r_retnet/commit/c2638d5d6fc8f0e94d20873020879f93b1d0392d).

nprisbrey commented on README.md:
````diff
@@ -1,116 +1,267 @@
-    batch_sizes = [32, 64, 128]
-    num_layers_options = [2, 4, 6]
+```bash
+./<some_script>.sh
````
Should we make this `<SOME_SCRIPT>` to be consistent with the convention of "Explicit Parameters" (introduced in #33) in this file and other files?

RLin8103 replied:
I will change <some_script> to <SOME_SCRIPT>.

RLin8103 replied:
I finished changing this.

nprisbrey commented on README.md:
```diff
@@ -1,116 +1,267 @@
+### Execution Workflow
 
-Assuming you have a training script `train.py`, you can run it with different hyperparameters.
+Once a YAML config file is prepared, you can pass this file into any script in the pipeline. Before you run any scripts, it is recommended to copy all needed scripts into the [user_scripts](./scripts/user_scripts/) folder and modify the scripts to point to the right config file.
```
The second sentence sounds weird due to saying "scripts" twice. Could this be fixed?

RLin8103 replied:
Maybe change "all needed scripts" to "all of them"?

RLin8103 replied:
I finished changing it.

nprisbrey commented on README.md:
````diff
@@ -1,116 +1,267 @@
 
 ```bash
-python train.py --learning-rate 0.001 --batch-size 64 --num-layers 4 --embed-dim 256
+./download_data.sh
+./train_tokenizer.sh
+./tokenize_data.sh
+
+# Replace this with the model that you want to run
+./retnet.sh
````
We never tell the user to `cd` into a scripts directory. Should these commands have their relative path added to the front (ex: `./scripts/user_scripts/download_data.sh`)?

RLin8103 replied:
I am adding `cd ./scripts/user_scripts` in the code.

RLin8103 replied:
I finished adding it.

nprisbrey commented on README.md:
````diff
@@ -1,116 +1,267 @@
 ```
 
-#### 2. **Using a Configuration File**:
+### Note on CUDA
+
+Nvidia GPUs are recommended to speed up computation, though a CPU can alternatively be used. Nvidia drivers and the CUDA toolkit will also need to be installed. To test if there is an Nvidia GPU available on your system, see instructions for [Linux](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#verify-you-have-a-cuda-capable-gpu) and [Windows](https://docs.nvidia.com/cuda/ ...
````
This section honestly seems a bit redundant; this is usually assumed knowledge in READMEs. I'm fine keeping this and resolving this comment, though I wanted to make it known that I found this unexpected.

RLin8103 replied:
Should I delete this section?

RLin8103 replied:
This section is deleted.

nprisbrey commented on README.md:
```diff
@@ -1,116 +1,267 @@
+
+We implement the grid search process as follows:
+
+- **Hyperparameters Tested:** Learning rates (`0.001`, `0.0005`, `0.0001`), embedding dimensions (`768`, `1024`, `1280`), feed-forward dimensions (`1024`, `2048`), heads (`4`, `8`), and sequence lengths (`256`, `512`) for a total of 72 unique combinations.
```
I'm not sure that it makes sense to include the specific hyperparameters directly into the README. Since these values can/probably will change, I think that we should simply refer to the script to find the actual values.

RLin8103 replied:
You mean refer to the default yaml file @nprisbrey?

tristancorncroissant replied:
@RLin8103, this was pre-yaml, and wouldn't be good for a yaml anyway because we'd be referening 144 different yamls. I think he means reference them to the actual grid_search.sh that we ran, but it isn't compatible with our current code base, so I think pointing users to it would be misleading. IMO, this is the easist way to communicate that info unless it's in the paper (appendix). Until it is, I feel like it's fine to leave it here.

nprisbrey commented on README.md:
```diff
@@ -1,116 +1,267 @@
+We implement the grid search process as follows:
+
+- **Hyperparameters Tested:** Learning rates (`0.001`, `0.0005`, `0.0001`), embedding dimensions (`768`, `1024`, `1280`), feed-forward dimensions (`1024`, `2048`), heads (`4`, `8`), and sequence lengths (`256`, `512`) for a total of 72 unique combinations.
+- **Evaluation Metric:** The models are compared based on their test loss, with a custom function `evaluate_models` indicating which model performed better.
```
Should we hint somewhere that the user can rewrite the `evaluate_models` method themselves for custom evaluation?

RLin8103 replied:
Should this be put in the grid search section, right behind this @nprisbrey?

RLin8103 replied:
I don't think that is necessary. With any library, you can just implement any function you need.

nprisbrey commented on README.md:
```diff
@@ -1,116 +1,267 @@
+- **Evaluation Metric:** The models are compared based on their test loss, with a custom function `evaluate_models` indicating which model performed better.
+- **Output:** Results are recorded in a CSV file, including each combination's average loss for both models, similarity scores, and training times.
+
+**Key Functions:**
+
+- `evaluate_models(model1, model2, model1_loss, model2_loss)`: Compares two models based on their test losses, returning a score that reflects their performance relative to each other.
+- `grid_search(config)`: Conducts the grid search over the predefined ranges of hyperparameters, training both RetNet and Transformer models with each combination and logging the results.
```
This is quite strange and very specific to have in the README. This should be a documentation string instead.

tristancorncroissant replied:
Removed

nprisbrey commented on README.md:
````diff
@@ -1,116 +1,267 @@
 
 ```bash
-python train.py --config config.yaml
+python3 scripts/user_scirpts/<YOUR_SCRIPT_NAME>.py configs/user_configs/<YOUR_CONFIG_HERE>.yaml
````
Just realized there's a typo in `user_scirpts`. Would it also be better to just run the `grid_search.sh` file instead?

tristancorncroissant replied:
Good suggestions, I implemented these edits.

nprisbrey commented on README.md:
```diff
@@ -1,116 +1,267 @@
+### PyTorch Lightning Integration
+
+PyTorch Lightning is leveraged in our project to streamline the training process of the RetNet and Transformer models, enabling efficient multi-core processing, easier scalability, and cleaner code by abstracting the boilerplate training loops. PyTorch Lightning's integration facilitates advanced functionalities like distributed training, automated logging, and checkpointing.
+
+**Key Advantages:**
+
+- **Simplified Training Loop**: By abstracting the complexity of the training loop, PyTorch Lightning allows us to focus on the model architecture and the experiment itself, rather than boilerplate code.
+- **Multi-Core and Distributed Training**: Lightning's built-in support for distributed training and multi-core processing significantly speeds up training times, allowing our models to leverage multiple GPUs seamlessly.
+- **Automated Checkpointing**: The custom checkpointing system, `CustomModelCheckpoint`, automatically saves model checkpoints and Hugging Face compatible weights during training, facilitating model preservation and reproducibility.
+- **Advanced Logging**: Integration with TensorBoard for detailed logging of training and validation metrics, helping in monitoring model performance and debugging.
+
+**Usage:**
+
+1. **Model Initialization**: Instantiate a model class with the desired configuration. The configuration should include model hyperparameters, training parameters, and dataset specifics.
+2. **Trainer Setup**: Configure a `Trainer` object from PyTorch Lightning, specifying training options such as the number of GPUs, distributed backend, and callbacks like model checkpointing.
+3. **Training Execution**: Use the `Trainer` to train the model by passing the model instance and the data module. The training process automatically handles device placement, distributed training, and logging.
+4. **Evaluation and Testing**: After training, use the `Trainer` for evaluating the model on a validation set and testing it on a test set, leveraging the best model checkpoint saved during training.
+
+### TorchScale Integration
+
+Our project leverages **Microsoft TorchScale** as the foundational framework to enhance the training and evaluation process of deep learning models, particularly focusing on the comparative analysis between RetNet and Transformer architectures. TorchScale is a robust library offering a suite of tools and utilities designed to optimize deep learning workflows, enabling scalable and efficient model ...
+
+**Core Contributions:**
+
+- **Efficiency and Scalability**: TorchScale provides critical functionalities for handling large-scale datasets and models, significantly improving training speed and efficiency without compromising accuracy or model complexity.
+- **Advanced Utilities**: The library includes a range of utilities for model evaluation, performance benchmarking, and hyperparameter tuning, facilitating a comprehensive analysis of model behaviors under various configurations.
+- **Integration with RetNet and Transformers**: Through TorchScale, we compare the performance of RetNet and Transformer models across different metrics, ensuring fair and rigorous evaluation standards.
+- **Enhanced Model Training**: TorchScale's support for distributed training and model optimization techniques allows us to train more complex models with larger datasets, pushing the boundaries of what's possible in our research.
+
+**Implementation Highlights:**
+
+Our project specifically benefits from TorchScale's:
+
+- **Distributed Training Support**: Utilizing TorchScale's distributed training capabilities to expedite the training process across multiple GPUs, enabling more extensive experimentation and faster iteration cycles.
+- **Performance Benchmarking Tools**: Leveraging built-in tools for benchmarking model performance, which has been crucial in the side-by-side comparison of RetNet and Transformer models, providing insights into their respective strengths and limitations.
+- **Hyperparameter Tuning and Model Evaluation**: Employing TorchScale's utilities for hyperparameter optimization and model evaluation to fine-tune our models for optimal performance, ensuring our comparisons are based on the best possible configurations of each architecture.
+
+**Usage in Our Project:**
+
+TorchScale's integration into our workflow has been very helpful, being utilized across various stages of model development from initial training to final evaluation. This enables us to conduct a thorough and nuanced comparison of RetNet and Transformer models.
```
This whole section should either be one paragraph or removed entirely. Why are we including a whole marketing campaign for PyTorch Lightning and TorchScale inside of the README? This adds little to no value to a user simply trying to get the code to run. This content is more like the content that should be included in the paper, which we link to and the user can read for this type of information.

RLin8103 replied:
@nprisbrey Maybe just one paragraph for each feature?

tristancorncroissant replied:
I moved this into the Overleaf.

nprisbrey commented on README.md:
```diff
@@ -1,116 +1,267 @@
-- Replace placeholders like `your_project_model` and `RetNetModel` with actual module and class names from your project.
-- These examples are meant to be adapted to your specific project setup and requirements.
-- The effectiveness of hyperparameter tuning can depend heavily on the specific nature of the dataset and task.
+## Others
```
What is this header even doing here?

RLin8103 replied:
Will delete that

RLin8103 replied:
This section is deleted.

nprisbrey commented on README.md:
````diff
@@ -1,116 +1,267 @@
+
+```
+@misc{eval-harness,
+  author       = {Gao, Leo and Tow, Jonathan and Abbasi, Baber and Biderman, Stella and Black, Sid and DiPofi, Anthony and Foster, Charles and Golding, Laurence and Hsu, Jeffrey and Le Noac'h, Alain and Li, Haonan and McDonell, Kyle and Muennighoff, Niklas and Ociepa, Chris and Phang, Jason and Reynolds, Laria and Schoelkopf, Hailey and Skowron, Aviya and Sutawika, Lintang and Tang, Eric and Thit ...
+  title        = {A framework for few-shot language model evaluation},
+  month        = 12,
+  year         = 2023,
+  publisher    = {Zenodo},
+  version      = {v0.4.0},
+  doi          = {10.5281/zenodo.10256836},
+  url          = {https://zenodo.org/records/10256836}
+}
+
+
 ```
+```
+@article{retnet,
+  author={Yutao Sun and Li Dong and Shaohan Huang and Shuming Ma and Yuqing Xia and Jilong Xue and Jianyong Wang and Furu Wei},
+  title     = {Retentive Network: A Successor to {Transformer} for Large Language Models},
+  journal   = {ArXiv},
+  volume    = {abs/2307.08621},
+  year      = {2023}
+}
 
-In `config.yaml`, you can have:
 
-```yaml
-learning_rate: 0.001
-batch_size: 64
-num_layers: 4
-embed_dim: 256
+```
+```
+@article{torchscale,
+  author    = {Shuming Ma and Hongyu Wang and Shaohan Huang and Wenhui Wang and Zewen Chi and Li Dong and Alon Benhaim and Barun Patra and Vishrav Chaudhary and Xia Song and Furu Wei},
+  title     = {{TorchScale}: {Transformers} at Scale},
+  journal   = {CoRR},
+  volume    = {abs/2211.13184},
+  year      = {2022}
+}
````
This whole section should be reduced and simplified. These citations should be included in the paper and deleted from the README. The purpose of this section is to give our citation once published for others to cite **us**. For now, this should be something simple stating that we're still waiting to be published and our citation will be given soon.

JayOrten replied:
I agree, most of this should be cut

RLin8103 replied:
I fixed this.

nprisbrey reviewed (changes requested):
In general, there is a lot of extra fluff that is not needed by a user in order to contribute or get up and running. Especially the second half of the README sounds like a marketing pitch instead of concise information that is relevant. I would argue that many of these sections be moved to the paper write-up in their current form and then either simplified or deleted. Detailed comments about sections are given below. Feel free to follow up with the comments individually if you think differently about something. See the GitHub Docs for [README expectations](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-readmes#about-readmes).

JayOrten commented on README.md:
````diff
@@ -1,116 +1,267 @@
+Ensure you have Python 3.11 installed. To check your current Python version, run:
+
+```bash
+python --version
````
Depending on our audience, this kind of thing is probably okay to omit. It's likely safe to assume that anyone who will be using this will already have some familiarity with basic python versioning. Would you agree @nprisbrey ?

RLin8103 replied:
I don't know. I just thought that know what version of Python to use is important since some features will not work in other versions. What is your thought on this @JayOrten, @nprisbrey, and @DrewGalbraith?

nprisbrey replied:
I find value in specifying what version of Python we're using, though it may be extra and unusual. I think that anything additional, such as explaining how to check the Python version, is equivalent to treating the reader as a child. Documentation and Google exist to help with this.

RLin8103 replied:
I deleted that instruction.

JayOrten commented on README.md:
```diff
@@ -1,116 +1,267 @@
+
+## Benchmarking and Results
+
+We use EleutherAI's open-source language model evaluation harness to empirically evaluate our models across a suite of different standardized tasks. These are largely 2-5 option multiple choice tasks, thus a randomly initialized language model will score 50-20% randomly guessing.
```
This is good, but this section should really just focus on how to run the benchmarks on a given model.

RLin8103 replied:
@DrewGalbraith Can you do this?

tristancorncroissant replied:
This is done.

Commit:
Delete Others Section

Commit:
Change Description For TorchScale RetNet and Transformer

RLin8103 committed:
Change <some_script> to <SOME_SCRIPT>

RLin8103 committed:
Delete Node On Cuda Section

RLin8103 committed:
Change all needed scripts to all of them

RLin8103 committed:
Add cd ./scripts/user_scripts in Execution Workflow

DrewGalbraith committed:
Mention optional environment instructions

RLin8103 committed:
Delete Citation References And The Section Simple

RLin8103 committed:
Delete Citation References And The Section Simple

RLin8103 committed:
Merge branch 'update_readme' of https://github.com/DRAGNLabs/301r_retnet into update_readme

RLin8103 committed:
Update README.md

Fix Typo out to our

RLin8103 committed:
Update README.md

Delete Check Python Version Code Instructions.

DrewGalbraith committed:
Fix typo and correct a file path

DrewGalbraith committed:
Add benchamrking instructions

DrewGalbraith committed:
Update README.md with yaml string warning

DrewGalbraith committed:
Remove extraneous text from setup instructions.

DrewGalbraith committed:
Cut general linux tutorials

DrewGalbraith committed:
General rewording and editing

DrewGalbraith committed:
Cut more extraeneous instructions

nprisbrey commented on scripts/job.sh:
@RLin8103, I'm confused by this file. Why are we introducing a new script that doesn't use the YAML config file?

RLin8103 replied:
scripts/job.sh is deleted.

nprisbrey commented on weights/2024-01-12-13:07:41_retnet_1853472/model_args.json:
Why are we uploading the weights file? How did this even get here? Don't we not even use this folder anymore? Did this get committed because new code was never pulled? @RLin8103

RLin8103 replied:
@nprisbrey, I don't know why and how this was added. I think I only modified the Readme.md. Do you know what happened @DrewGalbraith, @JayOrten?

RLin8103 replied:
This directory is deleted.

nprisbrey commented on README.md:
````diff
@@ -1,116 +1,165 @@
-    ...
-    scheduler.step()  # Update learning rate
-```
+*Note:* In most cases, YAML does not expect strings. Adding quotation marks around argurments in the config file can lead to unexpected errors.
````
Is this verified or only a one-off issue we saw on @jtappen1's setup?

RLin8103 replied:
@JayOrten, I don't think you need quotation marks in YAML, right?

tristancorncroissant replied:
No, it ruined mine as well. Jay said we yaml doesn't expect quotation marks.

nprisbrey commented on README.md:
````diff
@@ -1,116 +1,165 @@
-num_layers: 4
-embed_dim: 256
 ```
+mamba activate <YOUR_ENV_HERE> # Activate environment, if using one.
+cd /301r_retnet/slurm/
+cp run_eval.sh user_slurm/<NAME_OF_NEW_FILE>.sh  # Give your file a descriptive name, (e.g., 'retnet_40540_run_eval.sh')
````
Why do we all the sudden switch convention to making comments in-line instead of directly above the line they're referencing?

RLin8103 replied:
Formatting issue is resolved.

nprisbrey committed:
Enumerate Installation and Setup Steps

nprisbrey committed:
Reword Environment Setup Instructions

nprisbrey committed:
Tweak Relative Paths

nprisbrey committed:
Formatting Tweaks For README

nprisbrey reviewed (changes requested):
Looking a lot better! I've added some commits to address some things I saw, though I've added comments on things that I didn't change myself in this review. I feel that my comments in [this comment](https://github.com/DRAGNLabs/301r_retnet/pull/32/files/1b050282606156935a648140e0428cb079023134#r1488731537) and [this comment](https://github.com/DRAGNLabs/301r_retnet/pull/32/files/1b050282606156935a648140e0428cb079023134#r1488731880) still stand from the last review as well.

RLin8103 committed:
Update README.md

Resolve Code Comments Format Issues In Benchmark

nprisbrey commented on README.md:
````diff
@@ -1,116 +1,170 @@
 
-In `config.yaml`, you can have:
+```
+# Activate environment, if using one.
````
@RLin8103, the period should be removed to be consistent with all other comments.

RLin8103 replied:
The period is deleted.

RLin8103 committed:
Delete weights directory

RLin8103 committed:
Delete scripts/job.sh

RLin8103 committed:
Delete Period In Comment

nprisbrey committed:
Fix Path Error In README

nprisbrey committed:
Fix Typos In README

nprisbrey reviewed (approved):
Seems good to me. I'd be interested in @KimballNJardine's thoughts on string quotes in the configuration file. I'd definitely be interested in the other reviewer's thoughts and fresh eyes (@JayOrten, @JacksonSearle, and @KimballNJardine).

JayOrten commented:
> Seems good to me. I'd be interested in @KimballNJardine's thoughts on string quotes in the configuration file. I'd definitely be interested in the other reviewer's thoughts and fresh eyes (@JayOrten, @JacksonSearle, and @KimballNJardine).

See: https://stackoverflow.com/questions/19109912/yaml-do-i-need-quotes-for-strings-in-yaml

Obviously, it's a big discussion with no clear "right" answer. In general, strings in yaml files don't need quotes on strings. For our application and usage, and based on the weird problems John was having, I would say they are unnecessary for our use case. I'm not very concerned about being explicit considering the comments we have. 

However, I haven't taken the time to dig into the issue John had, so maybe there's a better fix and we can keep the quotes if everyone really cares about them.

KimballNJardine commented on README.md:
```diff
@@ -1,116 +1,170 @@
-Make sure to have Python=3.11
-When you use a .sh script, run this to grant permission "chmod +x retnet.sh" for the script you want to use
-Then do this "./retnet.sh" to run the script
+This project focuses on the comparison and analysis of the RetNet vs the Transformer architecture, utilizing Microsoft's TorchScale library for implementation. More information can be found in our paper [Are You Using Retentive Networks?](./301R_Retnet_Paper.pdf).
```
the RetNet vs the Transformer architecture
This phrasing looks strange.
perhaps "on the comparison and analysis of the RetNet and the Transformer architectures"?

tristancorncroissant replied:
Changed to "This project compares and analyzes the RetNet and the Transformer architectures."

KimballNJardine commented on README.md:
```diff
@@ -1,116 +1,170 @@
-- dataset_name : the name of the dataset on HF hub
-- dataset_subset : the subset to use
-- dataset_dir : the name of the folder to save the dataset in. Typically, you can just keep this the same as the dataset_name
+This project is built upon [Microsoft TorchScale](https://github.com/microsoft/torchscale), which provides basic implementations of the two architectures for our research. TorchScale provides a library of foundational architecture implementations for training Transformer-based deep learning models. We have leveraged its capabilities to develop our comparison between RetNets and Transformers.
```
between RetNet and Transformer
Perhaps
between the RetNet and Transformer architectures.

tristancorncroissant replied:
Distilled.

KimballNJardine commented on README.md:
```diff
@@ -1,116 +1,170 @@
+2. Run `download_data.sh` to fetch the necessary data.
+3. Execute `train_tokenizer.sh` to prepare the tokenizer.
+4. Use `tokenize_data.sh` for data tokenization.
+5. Finally, run `retnet.sh` or `transformer.sh`, depending on your project needs.
```
run `train_model.sh`, not `retnet.sh` or `transformer.sh`.

tristancorncroissant replied:
Yes, that's right. This is now done.

KimballNJardine commented on README.md:
```diff
@@ -1,116 +1,170 @@
-tuned_model = random_search_hyperparams()
-# Train and evaluate this model
+# Replace this with the file pertaining to the model that you want to run
+./retnet.sh
```
replace ./retnet.sh with ./train_model.sh

tristancorncroissant replied:
Done.

KimballNJardine commented on README.md:
```diff
@@ -1,116 +1,170 @@
+### 3: Script Execution
 
-Pseudo-code for implementing random search.
+Once a YAML config file is prepared, it can be passed into any script in the pipeline. Before you run any scripts, it is recommended to copy all of them into the [user_scripts](./scripts/user_scripts/) directory and modify the scripts to point to the right config file.
```
Look at the raw text below how to implement this in GitHub. I think this point should be emphasized.

> [!TIP]
> Once a YAML config file is prepared, it can be passed into any script in the pipeline. Before you run any scripts, it is recommended to copy all of them into the [user_scripts](./scripts/user_scripts/) directory and modify the scripts to point to the right config file.

tristancorncroissant replied:
I see this in the script, did you add it already?

KimballNJardine commented on README.md:
```diff
@@ -1,116 +1,170 @@
+
+We implement the grid search process as follows:
+
+- **Hyperparameters Tested:** Learning rates (`0.001`, `0.0005`, `0.0001`), embedding dimensions (`768`, `1024`, `1280`), feed-forward dimensions (`1024`, `2048`), heads (`4`, `8`), and sequence lengths (`256`, `512`) for a total of 72 unique combinations.
```
for a total of 72 unique combinations
--> change to 
for a total of 72 unique combinations per model architecture.

KimballNJardine replied:
Others commenting would be nice on the matter.

RLin8103 replied:
I will change that.

KimballNJardine commented on README.md:
````diff
@@ -1,116 +1,170 @@
+  - "hellaswag"
+  - "winogrande"
 ```
+Alternatively, you can use `tasks: '*'` to run all benchmarks in the suite. Then navigate to the `slurm/run_eval.sh`, copy the script, and substitute your yaml file for the placeholder. Finally, execute:
````
Where is the benchmark suite located? A path to it would be nice as a note.

tristancorncroissant replied:
Once downloaded, they land in `~/.cache/huggingface/datasets/`. I've added this path to the section.

KimballNJardine commented on README.md:
```diff
@@ -1,116 +1,170 @@
+
+## Benchmarking and Results
+
+We use EleutherAI's open-source language model evaluation harness to empirically evaluate our models across a suite of different NLP tasks. Run the evaluation suite as follows:
```
Noting that we run it locally on downloaded tests would be a good detail.

tristancorncroissant replied:
We've noted that they will need to be downloaded if not already stored locally at `~/.cache/huggingface/datasets/`; that should be enough for them to know they can be downloaded then run or if already downloaded, run without internet access, I think.

KimballNJardine commented on README.md:
```diff
@@ -1,116 +1,170 @@
+
+**BYU Office of Research Computing**: For providing the computational resources and support that were instrumental in conducting our experiments and analyses.
+
+**Authors of the Original RetNet Paper**: For their pioneering work in the field of deep learning, which inspired our research and provided a solid base for our explorations into RetNet and Transformers.
```
I would not say "pioneering work in the field of deep learning" --> that feels too generous, in that it is false. RetNet is not truly pioneering work in the field of deep learning.
Perhaps "For their work in novel encoder architectures"?

KimballNJardine commented on README.md:
```diff
@@ -1,116 +1,170 @@
-- Replace placeholders like `your_project_model` and `RetNetModel` with actual module and class names from your project.
-- These examples are meant to be adapted to your specific project setup and requirements.
-- The effectiveness of hyperparameter tuning can depend heavily on the specific nature of the dataset and task.
+Our paper is awaiting publication and our full citation will be given soon. Our current citations can be found on the citation section on our paper [Are You Using Retentive Networks?](./301R_Retnet_Paper.pdf).
```
Probably take this off. We should probably have either the citations on the readme itself, or something like the following:
> [!NOTE]
> Our paper is awaiting publication and our full citation will be given soon.

Adding on the citations here would be good without a link to the paper itself.

tristancorncroissant replied:
While I originally liked the idea of adding our citations into the README.md, the standard for paper repos seems to be add your own citation and let users find the rest in your bibliography (e.g. see anything from [AllenAI](https://github.com/allenai), Google's [BigBench](https://github.com/google/BIG-bench)).

tristancorncroissant replied:
I did delete the link to the pdf though.

tristancorncroissant replied:
How did you add that " (i) note" block?

KimballNJardine replied:
Remove the # below to see that.
> [!NOTE]
> Our paper is awaiting publication and our full citation will be given soon.

#> [!NOTE]
> Our paper is awaiting publication and our full citation will be given soon.

KimballNJardine replied:
https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

KimballNJardine commented on 301R_Retnet_Paper.pdf:
We probably want to remove this, as it provides a clear link back to our code and ourselves (you just have to look at the people committing to the repo itself to know who the coauthors are.)

KimballNJardine replied:
If someone else has a reason to keep the paper before peer review, please respond.

tristancorncroissant replied:
You are probably right. I've deleted it, though it's not scrubbed from the repo. I doubt that will be too important.

KimballNJardine reviewed (changes requested):
I have a concern about having the paper inside the repo itself, even if it is only the pdf. Also note other comments.

DrewGalbraith committed:
Delete 301R_Retnet_Paper.pdf

DrewGalbraith committed:
Enhance writing

DrewGalbraith committed:
edit writing

DrewGalbraith committed:
Update README.md with train_model.sh

DrewGalbraith committed:
Elimante links to deleted pdf

DrewGalbraith committed:
Add benchmark file path

KimballNJardine commented on README.md:
```diff
@@ -1,116 +1,168 @@
-This is a readme for the RETNET team
+# Are You Using Retentive Networks? (RetNets) 📝
```
Do we want the header to be the name of our paper? When I search up the paper name, this pops up in search results. I am not sure if this is anonymous enough.

tristancorncroissant replied:
I searched the whole title and a few variants in Google, but didn't get anything back on the 1st page of results in either Incognito nor a normal browsing session. I wonder if this is a cookies thing specific to those using the repo frequently by searching that name? Try looking it up in an Incognito browser.

KimballNJardine replied:
I get responses on DuckDuckGo and Bing, but not Google.

JayOrten replied:
![Screenshot_20240222-180032.png](https://github.com/DRAGNLabs/301r_retnet/assets/33641860/19b9a641-0f97-4998-8292-b43fe380a886)

tristancorncroissant replied:
Interesting. I still get nothing on Edge and Firefox. I guess we can change the title name for now. The paper title is more likely than not to undergo changes as well, given we're doing something of a survey now, but I'll put in a place holder.

KimballNJardine reviewed (changes requested):
Note comment on header.

DrewGalbraith committed:
Remove paper title from README.md

DrewGalbraith committed:
Remove paper title from README.md

RLin8103 committed:
Add Per Model Architecture After 72 Combination

RLin8103 committed:
Change Acknowledgement And Fix Minor Grammatic Issue

RLin8103 committed:
Add Note In Script Execution To Highlight

RLin8103 committed:
Change Note To Tip

RLin8103 committed:
Add Note To Citations And Change Wording

KimballNJardine commented:
> > Seems good to me. I'd be interested in @KimballNJardine's thoughts on string quotes in the configuration file. I'd definitely be interested in the other reviewer's thoughts and fresh eyes (@JayOrten, @JacksonSearle, and @KimballNJardine).
> 
> See: https://stackoverflow.com/questions/19109912/yaml-do-i-need-quotes-for-strings-in-yaml
> 
> Obviously, it's a big discussion with no clear "right" answer. In general, strings in yaml files don't need quotes on strings. For our application and usage, and based on the weird problems John was having, I would say they are unnecessary for our use case. I'm not very concerned about being explicit considering the comments we have.
> 
> However, I haven't taken the time to dig into the issue John had, so maybe there's a better fix and we can keep the quotes if everyone really cares about them.

I am fine without the quotes.

KimballNJardine reviewed (approved):
Looks good. Approved.

JayOrten reviewed (approved):
I approve
